File: 46-cost-intelligence-stack/python-sdk/basic_intelligence.py
"""
Cost Intelligence Stack - Basic Intelligence Example

Demonstrates how to:
- Get real-time cost intelligence insights
- View telemetry data and trends
- Access intelligence recommendations
"""

import os
from costkatana import CostKatana

client = CostKatana(api_key=os.getenv('COST_KATANA_API_KEY'))

def main():
    try:
        print('🧠 Cost Intelligence Stack - Basic Intelligence\n')

        # 1. Get current intelligence insights
        print('📊 Fetching intelligence insights...')
        insights = client.intelligence.get_insights(
            time_range='24h',
            include_recommendations=True
        )

        print('\n✅ Intelligence Insights:')
        print(f'Total Cost: ${insights["totalCost"]:.4f}')
        print(f'Total Requests: {insights["totalRequests"]}')
        print(f'Average Cost per Request: ${insights["avgCostPerRequest"]:.6f}')
        print(f'Anomalies Detected: {insights["anomaliesDetected"]}')

        # 2. Get telemetry trends
        print('\n📈 Fetching telemetry trends...')
        trends = client.intelligence.get_trends(
            time_range='7d',
            granularity='day',
            metrics=['cost', 'requests', 'latency']
        )

        print('\n✅ Trend Analysis:')
        print(f'Cost Trend: {trends["cost"]["direction"]} ({trends["cost"]["changePercent"]}%)')
        print(f'Request Trend: {trends["requests"]["direction"]} ({trends["requests"]["changePercent"]}%)')
        print(f'Latency Trend: {trends["latency"]["direction"]} ({trends["latency"]["changePercent"]}ms)')

        # 3. Get intelligence recommendations
        if insights.get('recommendations'):
            print('\n💡 Active Recommendations:')
            for index, rec in enumerate(insights['recommendations'], 1):
                print(f'\n{index}. {rec["title"]}')
                print(f'   Type: {rec["type"]}')
                print(f'   Potential Savings: ${rec["estimatedSavings"]:.4f}')
                print(f'   Priority: {rec["priority"]}')
                print(f'   Action: {rec["action"]}')

        # 4. Get layer-specific insights
        print('\n🏗️ Layer-Specific Insights:')
        layer_insights = client.intelligence.get_layer_insights()
        
        print('\nLayer 1 (Telemetry):')
        print(f'  Capture Rate: {layer_insights["telemetry"]["captureRate"]}%')
        print(f'  Streaming Active: {layer_insights["telemetry"]["streamingActive"]}')
        
        print('\nLayer 2 (Intelligence):')
        print(f'  Analysis Mode: {layer_insights["intelligence"]["mode"]}')
        print(f'  Confidence Score: {layer_insights["intelligence"]["confidenceScore"]}%')
        
        print('\nLayer 3 (Routing):')
        print(f'  Active Routes: {layer_insights["routing"]["activeRoutes"]}')
        print(f'  Optimization Rate: {layer_insights["routing"]["optimizationRate"]}%')
        
        print('\nLayer 4 (Enforcement):')
        print(f'  Blocked Requests: {layer_insights["enforcement"]["blockedRequests"]}')
        print(f'  Budget Utilization: {layer_insights["enforcement"]["budgetUtilization"]}%')
        
        print('\nLayer 5 (Caching):')
        print(f'  Cache Hit Rate: {layer_insights["caching"]["hitRate"]}%')
        print(f'  Cost Savings: ${layer_insights["caching"]["costSavings"]:.4f}')
        
        print('\nLayer 6 (Simulation):')
        print(f'  Active Simulations: {layer_insights["simulation"]["activeSimulations"]}')
        print(f'  Accuracy: {layer_insights["simulation"]["accuracy"]}%')

        print('\n✅ Intelligence query completed successfully!')

    except Exception as error:
        print(f'❌ Error: {error}')
        exit(1)

if __name__ == '__main__':
    main()

File: 46-cost-intelligence-stack/npm-package/basic-intelligence.ts
/**
 * Cost Intelligence Stack - Basic Intelligence Example
 * 
 * Demonstrates how to:
 * - Get real-time cost intelligence insights
 * - View telemetry data and trends
 * - Access intelligence recommendations
 */

import { CostKatana } from 'cost-katana';

const client = new CostKatana({
  apiKey: process.env.COST_KATANA_API_KEY!,
});

async function main() {
  try {
    console.log('🧠 Cost Intelligence Stack - Basic Intelligence\n');

    // 1. Get current intelligence insights
    console.log('📊 Fetching intelligence insights...');
    const insights = await client.intelligence.getInsights({
      timeRange: '24h',
      includeRecommendations: true,
    });

    console.log('\n✅ Intelligence Insights:');
    console.log(`Total Cost: $${insights.totalCost.toFixed(4)}`);
    console.log(`Total Requests: ${insights.totalRequests}`);
    console.log(`Average Cost per Request: $${insights.avgCostPerRequest.toFixed(6)}`);
    console.log(`Anomalies Detected: ${insights.anomaliesDetected}`);

    // 2. Get telemetry trends
    console.log('\n📈 Fetching telemetry trends...');
    const trends = await client.intelligence.getTrends({
      timeRange: '7d',
      granularity: 'day',
      metrics: ['cost', 'requests', 'latency'],
    });

    console.log(`\n✅ Trend Analysis:`);
    console.log(`Cost Trend: ${trends.cost.direction} (${trends.cost.changePercent}%)`);
    console.log(`Request Trend: ${trends.requests.direction} (${trends.requests.changePercent}%)`);
    console.log(`Latency Trend: ${trends.latency.direction} (${trends.latency.changePercent}ms)`);

    // 3. Get intelligence recommendations
    if (insights.recommendations && insights.recommendations.length > 0) {
      console.log('\n💡 Active Recommendations:');
      insights.recommendations.forEach((rec, index) => {
        console.log(`\n${index + 1}. ${rec.title}`);
        console.log(`   Type: ${rec.type}`);
        console.log(`   Potential Savings: $${rec.estimatedSavings.toFixed(4)}`);
        console.log(`   Priority: ${rec.priority}`);
        console.log(`   Action: ${rec.action}`);
      });
    }

    // 4. Get layer-specific insights
    console.log('\n🏗️ Layer-Specific Insights:');
    
    const layerInsights = await client.intelligence.getLayerInsights();
    
    console.log('\nLayer 1 (Telemetry):');
    console.log(`  Capture Rate: ${layerInsights.telemetry.captureRate}%`);
    console.log(`  Streaming Active: ${layerInsights.telemetry.streamingActive}`);
    
    console.log('\nLayer 2 (Intelligence):');
    console.log(`  Analysis Mode: ${layerInsights.intelligence.mode}`);
    console.log(`  Confidence Score: ${layerInsights.intelligence.confidenceScore}%`);
    
    console.log('\nLayer 3 (Routing):');
    console.log(`  Active Routes: ${layerInsights.routing.activeRoutes}`);
    console.log(`  Optimization Rate: ${layerInsights.routing.optimizationRate}%`);
    
    console.log('\nLayer 4 (Enforcement):');
    console.log(`  Blocked Requests: ${layerInsights.enforcement.blockedRequests}`);
    console.log(`  Budget Utilization: ${layerInsights.enforcement.budgetUtilization}%`);
    
    console.log('\nLayer 5 (Caching):');
    console.log(`  Cache Hit Rate: ${layerInsights.caching.hitRate}%`);
    console.log(`  Cost Savings: $${layerInsights.caching.costSavings.toFixed(4)}`);
    
    console.log('\nLayer 6 (Simulation):');
    console.log(`  Active Simulations: ${layerInsights.simulation.activeSimulations}`);
    console.log(`  Accuracy: ${layerInsights.simulation.accuracy}%`);

    console.log('\n✅ Intelligence query completed successfully!');

  } catch (error) {
    console.error('❌ Error:', error);
    process.exit(1);
  }
}

main();

